umh-core/pkg/fsm/agent/reconcile.go

```diff
@@ -0,0 +1,289 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	internal_fsm "github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/internal/fsm"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/backoff"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/fsm"
+	"github.com/united-manufacturing-hub/united-manufacturing-hub/umh-core/pkg/metrics"
+)
+
+// Constants for agent FSM
+const (
+	// AgentUpdateObservedStateTimeout defines the timeout for updating observed state
+	AgentUpdateObservedStateTimeout = time.Millisecond * 5
+)
+
+// Reconcile examines the AgentInstance and, in three steps:
+//  1. Check if a previous transition failed; if so, verify whether the backoff has elapsed.
+//  2. Detect any external changes (e.g., a new configuration or external signals).
+//  3. Attempt the required state transition by sending the appropriate event.
+//
+// This function is intended to be called repeatedly (e.g. in a periodic control loop).
+func (a *AgentInstance) Reconcile(ctx context.Context, tick uint64) (err error, reconciled bool) {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(ComponentAgentInstance, "reconcile", time.Since(start))
+		if err != nil {
+			a.baseFSMInstance.GetLogger().Errorf("Error reconciling agent: %v", err)
+			a.PrintState()
+			// Add metrics for error
+			metrics.IncErrorCount(ComponentAgentInstance, "agent")
+		}
+	}()
+
+	// Check if context is already cancelled
+	if ctx.Err() != nil {
+		return ctx.Err(), false
+	}
+
+	// Step 1: If there's a lastError, see if we've waited enough.
+	if a.baseFSMInstance.ShouldSkipReconcileBecauseOfError(tick) {
+		err := a.baseFSMInstance.GetBackoffError(tick)
+		a.baseFSMInstance.GetLogger().Debugf("Skipping reconcile for agent: %v", err)
+
+		// if it is a permanent error, start the removal process and reset the error
+		if backoff.IsPermanentFailureError(err) {
+			// If already in a terminal state with permanent error, force remove
+			if a.IsRemoved() || a.IsRemoving() || a.IsMonitoringStopped() {
+				a.baseFSMInstance.GetLogger().Errorf("Agent is already in a terminal state, force removing it")
+				// In a real implementation, there might be additional cleanup here
+				return err, false
+			} else {
+				a.baseFSMInstance.GetLogger().Errorf("Agent is not in a terminal state, resetting state and removing it")
+				a.baseFSMInstance.ResetState()
+				a.Remove(ctx)
+				return nil, false // Try to reconcile towards a monitoring_stopped/removed state
+			}
+		}
+		return nil, false
+	}
+
+	// Step 2: Detect external changes.
+	if err := a.reconcileExternalChanges(ctx, tick); err != nil {
+		a.baseFSMInstance.SetError(err, tick)
```

Comment:
check out the cases in the benthos reconcile loop, e.g.,

if errors.Is(err, context.DeadlineExceeded) {
				// Healthchecks occasionally take longer (sometimes up to 70ms),
				// resulting in context.DeadlineExceeded errors. In this case, we want to
				// mark the reconciliation as complete for this tick since we've likely
				// already consumed significant time. We return reconciled=true to prevent
				// further reconciliation attempts in the current tick.
				return nil, true // We don't want to return an error here, as this can happen in normal operations
			}